src/components/implementation/no_interface/llbooter/llbooter.c

```diff
@@ -108,24 +110,101 @@ boot_spd_end(struct cobj_header *h)
 	return sect->vaddr + round_up_to_page(sect->bytes);
 }
 
+struct cobj_symb *
+cobj_find_symb(const char *symbol_name, int cobj) {
+	unsigned int j = 0;
+	for (j = 0; j < hs[cobj]->nsymb; j++) {
+		struct cobj_symb *symb;
+
+		symb = cobj_symb_get(hs[cobj], j);
+		assert(symb);
+
+		if (!strcmp(symb->name, symbol_name) && symb->type == COBJ_SYMB_EXPORTED) {
+			// printc("found: %s, spdid: %d, addr: %x \n", symb->name, cobj, symb->vaddr);
+			return symb;
+		}
+				
+	}		
+	return NULL;
+}
+
+vaddr_t
+boot_find_invsymb_addr(struct cobj_symb *undefsymb) {
+	
+	unsigned int i;
+	vaddr_t symb_addr;	
+	struct cobj_symb *symb;
+	/* HACK is this the best way to do this? */
+	char jumper_symb[25];
+	strcpy(jumper_symb, undefsymb->name);
+	strcat(jumper_symb, "_inv");
+
+	for (i = 0; hs[i] != NULL; i++) {
+		symb = cobj_find_symb(jumper_symb, i);
+		if (symb) break;	
+	}
+
+	assert(symb);
+	symb_addr = symb->vaddr;
+
+	return symb_addr;
+}
+
 int
-boot_spd_symbs(struct cobj_header *h, spdid_t spdid, vaddr_t *comp_info)
+boot_link_symbs(struct cobj_header *h, spdid_t spdid)
 {
 	int i = 0;
+	struct cobj_symb *ipc_client_symb;
+	struct cobj_symb *symb;
+	vaddr_t symb_addr;
 
 	for (i = 0; i < (int)h->nsymb; i++) {
-		struct cobj_symb *symb;
 
 		symb = cobj_symb_get(h, i);
 		assert(symb);
-		if (COBJ_SYMB_UNDEF == symb->type) break;
+	
+		if (COBJ_SYMB_UNDEF == symb->type) {
+			ipc_client_symb = cobj_find_symb("SS_ipc_client_marshal_args", spdid-1);
+			symb_addr = ipc_client_symb->vaddr;
+		
+			/* create the user cap for the undef symb */
+			struct usr_inv_cap cap = (struct usr_inv_cap) {
```

Comment:
declarations at the top of the block (below the `if`)...

`cap = (struct ....) {...}`  will still work